### PR TITLE
887: Hash PasswordResetKey

### LINKED ISSUE #821: Exception when resetting password of a Nürnberg account on bayern.ehrenamtskarte.app
```
eb 24 07:11:07 entitlementcard.tuerantuer.org backend[18834]: [JettyServerThreadPool-20] WARN notprivacysafe.graphql.execution.SimpleDataFetcherExceptionHandler - Exception while fetching data (/result) : Collection is empty.
Feb 24 07:11:07 entitlementcard.tuerantuer.org backend[18834]: java.util.NoSuchElementException: Collection is empty.
Feb 24 07:11:07 entitlementcard.tuerantuer.org backend[18834]:         at kotlin.collections.CollectionsKt___CollectionsKt.single(_Collections.kt:594)
Feb 24 07:11:07 entitlementcard.tuerantuer.org backend[18834]:         at app.ehrenamtskarte.backend.auth.webservice.schema.ResetPasswordMutationService$sendResetMail$1.invoke(ResetPasswordMutationService.kt:25)
Feb 24 07:11:07 entitlementcard.tuerantuer.org backend[18834]:         at app.ehrenamtskarte.backend.auth.webservice.schema.ResetPasswordMutationService$sendResetMail$1.invoke(ResetPasswordMutationService.kt:22)
Feb 24 07:11:07 entitlementcard.tuerantuer.org backend[18834]:         at org.jetbrains.exposed.sql.transactions.ThreadLocalTransactionManagerKt.inTopLevelTransaction$run(ThreadLocalTransactionManager.kt:215)
Feb 24 07:11:07 entitlementcard.tuerantuer.org backend[18834]:         at org.jetbrains.exposed.sql.transactions.ThreadLocalTransactionManagerKt.access$inTopLevelTransaction$run(ThreadLocalTransactionManager.kt:1)
Feb 24 07:11:07 entitlementcard.tuerantuer.org backend[18834]:         at org.jetbrains.exposed.sql.transactions.ThreadLocalTransactionManagerKt$inTopLevelTransaction$1.invoke(ThreadLocalTransactionManager.kt:241)
Feb 24 07:11:07 entitlementcard.tuerantuer.org backend[18834]:         at org.jetbrains.exposed.sql.transactions.ThreadLocalTransactionManagerKt.keepAndRestoreTransactionRefAfterRun(ThreadLocalTransactionManager.kt:249)
Feb 24 07:11:07 entitlementcard.tuerantuer.org backend[18834]:         at org.jetbrains.exposed.sql.transactions.ThreadLocalTransactionManagerKt.inTopLevelTransaction(ThreadLocalTransactionManager.kt:240)
Feb 24 07:11:07 entitlementcard.tuerantuer.org backend[18834]:         at org.jetbrains.exposed.sql.transactions.ThreadLocalTransactionManagerKt$transaction$1.invoke(ThreadLocalTransactionManager.kt:190)
Feb 24 07:11:07 entitlementcard.tuerantuer.org backend[18834]:         at org.jetbrains.exposed.sql.transactions.ThreadLocalTransactionManagerKt.keepAndRestoreTransactionRefAfterRun(ThreadLocalTransactionManager.kt:249)
Feb 24 07:11:07 entitlementcard.tuerantuer.org backend[18834]:         at org.jetbrains.exposed.sql.transactions.ThreadLocalTransactionManagerKt.transaction(ThreadLocalTransactionManager.kt:160)
Feb 24 07:11:07 entitlementcard.tuerantuer.org backend[18834]:         at org.jetbrains.exposed.sql.transactions.ThreadLocalTransactionManagerKt.transaction(ThreadLocalTransactionManager.kt:146)
Feb 24 07:11:07 entitlementcard.tuerantuer.org backend[18834]:         at org.jetbrains.exposed.sql.transactions.ThreadLocalTransactionManagerKt.transaction$default(ThreadLocalTransactionManager.kt:145)
Feb 24 07:11:07 entitlementcard.tuerantuer.org backend[18834]:         at app.ehrenamtskarte.backend.auth.webservice.schema.ResetPasswordMutationService.sendResetMail(ResetPasswordMutationService.kt:22)
```
- Check first if user exists in the project and return exception "User does not exist"

### PR DESCRIPTION
**Pull request by sarahsporck**
### Short description

To ensure that the password reset keys are safe even if the DB is somehow exposed we only want to store the hash of the password reset key.

### Proposed changes

- migrate db 1) change datatype to bytea 2) change column name
- add hashing logic
- change referential equality check on project to structural equality check this actually fixes #821 
- correctly handle runtime errors in frontend

### Side effects

- existing passwordResetKeys become invalid

### Resolved issues

Fixes: #887

### CODE REVIEW

File: administration/src/errors/getMessageFromApolloError.tsx [hunk 1]
```diff
@@ -16,7 +16,7 @@ const getMessageFromApolloError = (error: ApolloError): GraphQLErrorMessage => {
   }
 
   const codesEqual = error.graphQLErrors.every(
-    (value, index, array) => value.extensions.code === array[0].extensions.code
+    (value, index, array) => value.extensions?.code === array[0].extensions?.code
```
Comment (sarahsporck) [inline administration/src/errors/getMessageFromApolloError.tsx:19]: If there are runtime errors in the backend there will be no code in the return message (even though an error occured)

File: backend/src/main/kotlin/app/ehrenamtskarte/backend/auth/webservice/schema/ResetPasswordMutationService.kt [hunk 2]
```diff
@@ -47,19 +48,20 @@ class ResetPasswordMutationService {
...
+        if (!backendConfig.projects.any { it.id == project }) throw ProjectNotFoundException(project)
         transaction {
             val user = Administrators.innerJoin(Projects).slice(Administrators.columns)
                 .select((Projects.project eq project) and (LowerCase(Administrators.email) eq email.lowercase()) and (Administrators.deleted eq false))
                 .singleOrNull()?.let { AdministratorEntity.wrapRow(it) }
 
+            val passwordResetKeyHash = user?.passwordResetKeyHash
             // We don't send error messages for empty collection to the user to avoid scraping of mail addresses
-            if (user === null) {
+            if (user === null || passwordResetKeyHash === null) {
```
Comment (sarahsporck) [inline backend/src/main/kotlin/app/ehrenamtskarte/backend/auth/webservice/schema/ResetPasswordMutationService.kt:59]: check is required for proper typings

File: backend/src/main/kotlin/app/ehrenamtskarte/backend/auth/database/PasswordCrypto.kt [hunk 3]
```diff
@@ -10,4 +11,10 @@ object PasswordCrypto {
 
     fun verifyPassword(password: String, hash: ByteArray) =
         BCrypt.verifyer().verify(password.toCharArray(), hash).verified
+
+    fun hashPasswordResetKey(passwordResetKey: String): ByteArray =
+        BCrypt.with(LongPasswordStrategies.none()).hash(cost, passwordResetKey.toCharArray())
```
[the remark this answers is not in the document]
Comment (michael-markl) [reply backend/src/main/kotlin/app/ehrenamtskarte/backend/auth/database/PasswordCrypto.kt:16]: I think we should probably not use bcrypt for the reset tokens. Bcrypt is very expensive, and good for short/bad/low-entropy passwords.
Our keys have a high entropy (as they're 64 random bytes). Thus, I think we can simply use SHA-256. This should suffice for any preimage attacks.
@maxammann any comment from our security expert? :D

[the remark this answers is not in the document]
Comment (maxammann) [reply backend/src/main/kotlin/app/ehrenamtskarte/backend/auth/database/PasswordCrypto.kt:16]: Yes SHA256 should be enough. The chances of generating 512bits that result in the target hash  are astronomically low.

But that is only true if the randomness of the bytes is good enough.

[the remark this answers is not in the document]
Comment (maxammann) [reply backend/src/main/kotlin/app/ehrenamtskarte/backend/auth/database/PasswordCrypto.kt:16]: SecureRandom was used so this should be fine.

**Review by michael-markl — APPROVED** [review]
@sarahsporck I've went ahead and replaced bcrypt with SHA-256.

**Comment by michael-markl** [comment]
@ztefanie @f1sh1918 Feel free to review this once more before merging.

**Pull request merged**